### PR TITLE
Allow setting multiple path in KUBECONFIG

### PR DESCRIPTION
**Pull request by qiujian16**
Change to user default loading rules that will automatically load config in KUBECONFIG environment variable.

fix https://github.com/istio/istio/issues/4081

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/client.go [hunk 1]
```diff
@@ -70,14 +70,25 @@ func ResolveConfig(kubeconfig string) (string, error) {
 
 // CreateInterface is a helper function to create Kubernetes interface
 func CreateInterface(kubeconfig string) (*rest.Config, kubernetes.Interface, error) {
```
Comment (baodongli) [inline pilot/pkg/serviceregistry/kube/client.go]: This is used by both the pilot server and istioctl. I believe that the issue you are trying to fix is from istioctl. I don't think pilot is ready to support this yet.

Comment (qiujian16) [reply pilot/pkg/serviceregistry/kube/client.go]: @baodongli I do not think there is difference between pilot server and istioctl...if pilot server does not specify kubeconfig, in cluster config will be used. This is to ensure that the kubeconfig loaded by istioctl is the same as what is loaded by kubernetes client.

File: pilot/pkg/serviceregistry/kube/client.go [hunk 2]
```diff
@@ -70,14 +70,25 @@ func ResolveConfig(kubeconfig string) (string, error) {
...
-	if err != nil {
-		return nil, nil, err
-	}
+	var config *rest.Config
+	var err error
 
-	config, err := clientcmd.BuildConfigFromFlags("", kube)
-	if err != nil {
-		return nil, nil, err
+	kubeconfigenv := os.Getenv("KUBECONFIG")
```
Comment (gyliu513) [inline pilot/pkg/serviceregistry/kube/client.go]: Can you put this to a helper function name as `ResolveMultiPathConfig` or some others

File: pilot/pkg/serviceregistry/kube/client.go [hunk 3]
```diff
@@ -70,14 +70,25 @@ func ResolveConfig(kubeconfig string) (string, error) {
...
+	if kubeconfig != "" || kubeconfigenv != "" {
+		loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+		loadingRules.ExplicitPath = kubeconfig
+		config, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
```
Comment (gyliu513) [inline pilot/pkg/serviceregistry/kube/client.go]: Add a new test for this in controller_test.go？

Comment (qiujian16) [reply pilot/pkg/serviceregistry/kube/client.go]: done

File: pilot/pkg/serviceregistry/kube/client.go (hunk 2 above)
Comment (qiujian16) [reply pilot/pkg/serviceregistry/kube/client.go]: done

File: pilot/pkg/serviceregistry/kube/client.go [hunk 4]
```diff
@@ -68,16 +68,39 @@ func ResolveConfig(kubeconfig string) (string, error) {
...
+	config, err := LoadConfigFromMultiPath(kubeconfig)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	config, err := clientcmd.BuildConfigFromFlags("", kube)
-	if err != nil {
-		return nil, nil, err
+	// switch to use in-cluster config
+	if config == nil {
```
Comment (gyliu513) [inline pilot/pkg/serviceregistry/kube/client.go]: Why not put the `in-cluster` logic to `LoadConfigFromMultiPath`?

File: pilot/pkg/serviceregistry/kube/client_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2017 Istio Authors
```
Comment (gyliu513) [inline pilot/pkg/serviceregistry/kube/client_test.go]: +1 to create a new test file.

**Pull request closed without merging**